{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This Notebook is used to get additional information about the dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Including the following items:\n",
    "1. The coordinates of each transcript in the dataset\n",
    "2. The CDS, UTR, exone regions for each transcript in the dataset.\n",
    "3. Easy to fetch the data only in a few steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Holds classes and functions for reading biological data (FASTA, BED, BigWig, etc.). \n",
    "\n",
    "   In general, the classes and functions in this module are only used for constructing the TFDS dataset.\n",
    "   Users which want to train a model on an existing dataset, or use a pre-trained model, do not need to \n",
    "   use this module. Consequently, this module is not imported in the __init__.py file and libraries such \n",
    "   as pyBigWig are not required for the installation of parnet. \n",
    "\"\"\"\n",
    "\n",
    "import math\n",
    "\n",
    "import yaml\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "import tensorflow_datasets as tfds\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "# %%\n",
    "# map bases to their integer representation\n",
    "base2int = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n",
    "\n",
    "# map bases to their complement\n",
    "baseComplement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "\n",
    "def reverse_complement(dna_string):\n",
    "    \"\"\"Returns the reverse-complement for a DNA string.\"\"\"\n",
    "\n",
    "    complement = [baseComplement.get(base, 'N') for base in dna_string]\n",
    "    reversed_complement = reversed(complement)\n",
    "    return ''.join(list(reversed_complement))\n",
    "\n",
    "def sequence2int(sequence, mapping=base2int):\n",
    "    \"\"\"Converts a DNA sequence to a list of integers.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): DNA sequence.\n",
    "        mapping (dict, optional): Character to integer mapping. Defaults to base2int.\n",
    "\n",
    "    Returns:\n",
    "        list: List of integers.\n",
    "    \"\"\"\n",
    "    return [mapping.get(base, 999) for base in sequence] # TODO: 999 is a hack, should be replaced by something else (e.g. -1?). \n",
    "\n",
    "def sequence2onehot(sequence, mapping=base2int):\n",
    "    \"\"\"Converts a DNA sequence to a one-hot encoded tf.Tensor.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): DNA sequence.\n",
    "        mapping (dict, optional): Character to integer mapping. Defaults to base2int.\n",
    "\n",
    "    Returns:\n",
    "        tf.Tensor: One-hot encoded sequence.\n",
    "    \"\"\"\n",
    "    return tf.one_hot(sequence2int(sequence, mapping), depth=4) # Remove tensorflow dependency, ideally this should use just numpy. \n",
    "\n",
    "def mask_noncanonical_bases(sequence):\n",
    "    \"\"\"Masks non-canonical bases (anything except A, C, G, T or whatever mapping is provided) with N. \n",
    "\n",
    "    Args:\n",
    "        sequence (str): DNA sequence.\n",
    "\n",
    "    Returns:\n",
    "        str: Masked DNA sequence.\n",
    "    \"\"\"\n",
    "    return ''.join([base if base in base2int else 'N' for base in sequence])\n",
    "\n",
    "# %%\n",
    "class Fasta():\n",
    "    def __init__(self, filepath, mask_noncanonical_bases=True) -> None:\n",
    "        \"\"\"\n",
    "        Initialize a Fasta object.\n",
    "\n",
    "        Args:\n",
    "            filepath (str): The path to the FASTA file.\n",
    "            mask_noncanonical_bases (bool, optional): Whether to mask non-canonical bases with 'N'. Defaults to True.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            import pysam\n",
    "        except:\n",
    "            raise ModuleNotFoundError('Please install pyBigWig. See https://github.com/pysam-developers/pysam')\n",
    "\n",
    "        self.mask_noncanonical_bases = mask_noncanonical_bases\n",
    "        self._fasta = pysam.FastaFile(filepath)\n",
    "    \n",
    "    def fetch(self, chrom, start, end, strand='+', **kwargs):\n",
    "        \"\"\"\n",
    "        Fetch the sequence from the FASTA file.\n",
    "\n",
    "        Args:\n",
    "            chrom (str): The chromosome name.\n",
    "            start (int): The start position.\n",
    "            end (int): The end position.\n",
    "            strand (str, optional): The strand of the sequence. Defaults to '+'.\n",
    "            **kwargs: Additional keyword arguments.\n",
    "\n",
    "        Returns:\n",
    "            np.ndarray: The one-hot encoded sequence as a numpy array.\n",
    "        \"\"\"\n",
    "        sequence = self._fasta.fetch(chrom, start, end).upper()\n",
    "\n",
    "        if self.mask_noncanonical_bases:\n",
    "            # Everything except A, C, G, T will be mapped to N\n",
    "            sequence = mask_noncanonical_bases(sequence)\n",
    "\n",
    "        # Reverse complement if necessary\n",
    "        if strand == '+':\n",
    "            pass\n",
    "        elif strand == '-':\n",
    "            sequence = ''.join(reverse_complement(sequence))\n",
    "        else:\n",
    "            raise ValueError(f'Unknown strand: {strand}')\n",
    "        \n",
    "        # Convert to one-hot encoded numpy array. We assume that the alphabet is fairly small (usually 4 or 5 bases). \n",
    "        return np.array(sequence2onehot(sequence), dtype=np.int8)\n",
    "\n",
    "    def __call__(self, *args, **kwargs):\n",
    "        return self.fetch(*args, **kwargs)\n",
    "\n",
    "# %%\n",
    "class Bed():\n",
    "    def __init__(self, filepath) -> None:\n",
    "        \n",
    "        self.bed_df = pd.read_csv(filepath, sep='\\t', header=None)\n",
    "        self.bed_df.columns = ['chrom', 'start', 'end', 'name', 'score', 'strand'] + [str(i) for i in range(6, len(self.bed_df.columns))]\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.bed_df)\n",
    "\n",
    "    def __iter__(self):\n",
    "        for i in range(0, len(self.bed_df)):\n",
    "            yield self.bed_df.loc[i].to_dict()\n",
    "\n",
    "# %%\n",
    "def nan_to_zero(x):\n",
    "    \"\"\"Replaces nan's with zeros.\"\"\"\n",
    "    return 0 if math.isnan(x) else x\n",
    "\n",
    "class BigWig():\n",
    "    def __init__(self, bigwig_filepath) -> None:\n",
    "        try:\n",
    "            import pyBigWig #TODO: For now I prefer to have imports at the top of the file. \n",
    "        except ModuleNotFoundError:\n",
    "            raise ModuleNotFoundError('Please install pyBigWig. See https://github.com/deeptools/pyBigWig')\n",
    "\n",
    "        self._bigWig = pyBigWig.open(bigwig_filepath)\n",
    "    \n",
    "    def values(self, chrom, start, end, **kwargs):\n",
    "        chrom, start, end = str(chrom), int(start), int(end) # not sure why this is needed, it worked locally with numpy.int32\n",
    "        values = self._bigWig.values(chrom, start, end)\n",
    "        values = [nan_to_zero(v) for v in values]        \n",
    "        return np.array(values, dtype=np.float32)\n",
    "    \n",
    "    def __call__(self, *args, **kwargs):\n",
    "        return self.values(*args, **kwargs)\n",
    "\n",
    "class StrandedBigWig():\n",
    "    def __init__(self, bigwig_plus, bigwig_minus, reverse_minus=True) -> None:\n",
    "        self._bigWig_plus = BigWig(bigwig_plus)\n",
    "        self._bigWig_minus = BigWig(bigwig_minus)\n",
    "        self.reverse_minus = reverse_minus\n",
    "\n",
    "    def values(self, chrom, start, end, strand='+', **kwargs):\n",
    "        \"\"\"Returns values for a given range and strand. \n",
    "        \n",
    "        Args:\n",
    "            chrom  (str): Chromosome (chr1, chr2, ...)\n",
    "            start  (int): 0-based start position\n",
    "            end    (int): 0-based end position\n",
    "            strand (str): Strand ('+' or '-')\n",
    "            \n",
    "        Returns:\n",
    "            numpy.array: Numpy array of shape (end-start, )\n",
    "        \"\"\"\n",
    "\n",
    "        if strand == '+':\n",
    "            bigWig = self._bigWig_plus\n",
    "        elif strand == '-':\n",
    "            bigWig = self._bigWig_minus\n",
    "        else:\n",
    "            raise ValueError(f'Unexpected strand: {strand}')\n",
    "\n",
    "        values = bigWig.values(chrom, start, end)\n",
    "\n",
    "        if strand == '-' and self.reverse_minus:\n",
    "            values = values[::-1]\n",
    "\n",
    "        return values\n",
    "    \n",
    "    def __call__(self, *args, **kwargs):\n",
    "        return self.values(*args, **kwargs)\n",
    "    \n",
    "# %%\n",
    "class DataSpec:\n",
    "    \"\"\"Specifies the data layout and assembles file-connectors to generate samples. \n",
    "\n",
    "       Inputs and outputs are specified in a YAML file. The YAML file should have the following structure:\n",
    "\n",
    "        inputs:\n",
    "            sequence: path/to/fasta\n",
    "        outputs:\n",
    "            Task_1:\n",
    "                eCLIP:\n",
    "                    - path/to/eclip/counts/forward\n",
    "                    - path/to/eclip/counts/reverse\n",
    "                control:\n",
    "                    - path/to/control/counts/forward\n",
    "                    - path/to/control/counts/reverse\n",
    "            ...\n",
    "            Task_N:\n",
    "                eCLIP:\n",
    "                    - path/to/eclip/counts/forward\n",
    "                    - path/to/eclip/counts/reverse\n",
    "                control:\n",
    "                    - path/to/control/counts/forward\n",
    "                    - path/to/control/counts/reverse\n",
    "\n",
    "        The YAML specification is then used to initialize Fasta and StrandedBigWig file-connectors, from which \n",
    "        samples can be fetch in the specified structure. For efficiency (storage and during training), the \n",
    "        extracted 1D bigWig stracks are stacked to two tensors for eCLIP and control, both of the shape (n_tasks, n_positions). \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, dataspec_yml, control=False):\n",
    "        self.control = control\n",
    "\n",
    "        # parse YAML dataspec\n",
    "        with open(dataspec_yml) as f:\n",
    "            self._dataspec = yaml.load(f, yaml.FullLoader)\n",
    "        self.tasks = self._dataspec['outputs'].keys()\n",
    "\n",
    "        # initialize fasta-file connector\n",
    "        self._dataspec['inputs']['sequence'] = Fasta(self._dataspec['inputs']['sequence'])\n",
    "\n",
    "        # initialize eCLIP bigWig-file connectors, one for each task\n",
    "        for task in self._dataspec['outputs']:\n",
    "            self._dataspec['outputs'][task]['eCLIP'] = StrandedBigWig(*self._dataspec['outputs'][task]['eCLIP'])\n",
    "\n",
    "        # if control counts are available, initialize bigWig-file connectors for controls\n",
    "        if control:\n",
    "            for task in self._dataspec['outputs']:\n",
    "                self._dataspec['outputs'][task]['control'] = StrandedBigWig(*self._dataspec['outputs'][task]['control'])\n",
    "    \n",
    "    @property\n",
    "    def tf_signature(self):\n",
    "        \"\"\"Returns the features of the data in tf.TensorSpec format (required for tf.data.Dataset.from_generator). \n",
    "\n",
    "        Returns:\n",
    "            dict: Nested dictionary of tf.TensorSpecs.\n",
    "        \"\"\"\n",
    "\n",
    "        signature = {\n",
    "            'meta': {\n",
    "                'name': tf.TensorSpec(shape = (), dtype=tf.string),\n",
    "            },\n",
    "            'inputs': {\n",
    "                'sequence': tf.TensorSpec(shape = (None, 4), dtype=tf.int8),\n",
    "            },\n",
    "            'outputs': {\n",
    "                'eCLIP': tf.TensorSpec(shape = (None, None), dtype=tf.float32),\n",
    "            },\n",
    "        }\n",
    "        if self.control:\n",
    "            signature['outputs']['control'] = tf.TensorSpec(shape = (None, None), dtype=tf.float32)\n",
    "        return signature\n",
    "    \n",
    "    @property\n",
    "    def tfds_features(self):\n",
    "        \"\"\"Returns the features of the data in tfds.features.FeaturesDict format.\n",
    "\n",
    "        Returns:\n",
    "            tfds.features.FeatureDict: TFDS features describing the data. \n",
    "        \"\"\"\n",
    "\n",
    "        features = {\n",
    "            'meta': {\n",
    "                'name': tfds.features.Tensor(shape=(), dtype=tf.string),\n",
    "            },\n",
    "            'inputs': {\n",
    "                'sequence': tfds.features.Tensor(shape=(None, None), dtype=tf.int8, encoding='zlib'),\n",
    "            },\n",
    "            'outputs': {\n",
    "                'eCLIP': tfds.features.Tensor(shape=(None, None), dtype=tf.float32, encoding='zlib'),\n",
    "            },\n",
    "        }\n",
    "        if self.control:\n",
    "            features['outputs']['control'] = tfds.features.Tensor(shape=(None, None), dtype=tf.float32, encoding='zlib')\n",
    "        features = tfds.features.FeaturesDict(features)\n",
    "        return features\n",
    "    \n",
    "\n",
    "    def fetch_sample(self, chrom, start, end, strand, target_size):\n",
    "        \"\"\"Fetches a sample from the data. \n",
    "\n",
    "        Samples are specified by chrom, start, end and strand. The sequence may also be padded to target_size. \n",
    "        Fetching a sample returns a dictionary with the following structure:\n",
    "\n",
    "        {\n",
    "            'meta': {\n",
    "                'name': Tensor(shape=(), dtype=string),\n",
    "            },\n",
    "            'inputs': {\n",
    "                'sequence': Tensor(shape=(target_size, 4), dtype=int8),\n",
    "            },\n",
    "            'outputs': {\n",
    "                'eCLIP': Tensor(shape=(n_tasks, target_size), dtype=float32),\n",
    "                'control': Tensor(shape=(n_tasks, target_size), dtype=float32),\n",
    "            },\n",
    "        }\n",
    "\n",
    "        Args:\n",
    "            chrom (str): Chromosome. \n",
    "            start (int): Start position (closed).\n",
    "            end (int): End position (open).\n",
    "            strand (int): Strand ('+' or '-').\n",
    "            target_size (int): Target size of the sequence. If the sequence is shorter than target_size, it will be padded.\n",
    "\n",
    "        Returns:\n",
    "            dict: Dictionary containing the sample tensors. \n",
    "        \"\"\"\n",
    "\n",
    "        sample = {'meta': {}, 'inputs': {}, 'outputs': {}}\n",
    "\n",
    "        # prepare padding\n",
    "        # TODO: Move this to a separate function. \n",
    "        size = end - start\n",
    "        padding_left = int(np.ceil((target_size-size)/2))\n",
    "        padding_right = int(np.floor((target_size-size)/2))\n",
    "\n",
    "\n",
    "        sample['inputs']['sequence'] = self._dataspec['inputs']['sequence'](chrom, start, end, strand)\n",
    "        sample['inputs']['sequence'] = tf.pad(sample['inputs']['sequence'], paddings=[[padding_left, padding_right], [0, 0]])\n",
    "\n",
    "        sample['outputs']['eCLIP'] = np.stack([self._dataspec['outputs'][task]['eCLIP'](chrom, start, end, strand) for task in self.tasks])\n",
    "        sample['outputs']['eCLIP'] = tf.pad(sample['outputs']['eCLIP'], paddings=[[0, 0], [padding_left, padding_right]])\n",
    "        if self.control:\n",
    "            sample['outputs']['control'] = np.stack([self._dataspec['outputs'][task]['control'](chrom, start, end, strand) for task in self.tasks])\n",
    "            sample['outputs']['control'] = tf.pad(sample['outputs']['control'], paddings=[[0, 0], [padding_left, padding_right]])\n",
    "\n",
    "        # meta/name\n",
    "        sample['meta']['name'] = tf.constant(f'{chrom}:{start}-{end}:{strand}', dtype=tf.string)\n",
    "\n",
    "        # assert padding\n",
    "        assert target_size == sample['outputs']['eCLIP'].shape[1] == sample['inputs']['sequence'].shape[0]\n",
    "\n",
    "        return sample\n",
    "    \n",
    "    def __call__(self, *args, **kwargs):\n",
    "        return self.fetch_sample(*args, **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta = \"/home/sxr280/DeepLocRNA/DeepLocRNA/data/allRNA/allRNA_all_human_data_seq_mergedm3locall2_pooled_deduplicated3_filtermilncsnsno.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "fasta_file = pysam.FastaFile(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> 010000000,Gene_ID:NCBI:1305,Refseq_ID:NM_001130103,Species:Homo\n",
      "ATCCGATCCGTGCCGCCGCAGGGAGTGTGTCAAGTTACAGAGGCGCCGGAATCGGCCCCTGCGCTCCTCGCCAGCCGCCACGACCCACCTCTGCCCATGGGGCCCTCCGTGTGCGCCCCTTCGCCCGGGGACTGAAACTGACTGGCCCGGGAGACACGAGGCGCCCAGAAGGACTGACAGCGCGGCACCAACTGCTCTGCAGACACTTGAAGGGAAAGACTGGGCGGAGAGAAGGAGAGCCGGTCAGATTCCCCTAACTTTCCTGGACTTGGAACGTTCTTCGAAATAACTTTTTTCTCACCTAGGTGTACCCCAATTACCGCTGGTTGTGCTTTTTCGGCACTTCCTCTCCTACTGCTAATTTTTCCGTCCTCTTTGCCGGGAGCAGCGGAAAGGGACGTTTTCCAGCGATACAAGCCCTTTCCCCCTGCCCCGCAGTTTGGATAGAGCCTTTTGGCAGCGGCTGTCGCCTTTATTTATTCTATTTATTTATTTATTGGTTCTCAAGACGCGAGAGGATGGTAGCGGAGCGCACCCACAAAGCGGCAGCCACCGGTGCCCGCGGCCCTGGGGAGTTGGGCGCGCCCGGGACGGTGGCTCTGGTGGCGGCGCGGGCGGAGCGCGGCGCACGGCTGCCGAGTCCAGGGTCGTGCGGGCTGCTGACGCTGGCCCTCTGCTCGCTGGCACTCAGCCTGCTCGCCCACTTTCGGACGGCCGAGCTGCAGGCCCGGGTGCTGCGCCTGGAAGCGGAGCGCGGGGAGCAGCAAATGGAGACGGCTATTTTGGGACGAGTCAATCAACTGCTGGACGAGAAATGGAAGCTCCACTCAAGGAGGCGCCGGGAGGCCCCAAAGACATCTCCAGGATGTAACTGCCCACCAGGACCTCCTGGTCCCACTGGAAGACCCGGACTCCCAGGGGACAAAGGTGCCATTGGGATGCCTGGACGTGTGGGGTCCCCCGGAGACGCTGGGCTGTCCATCATTGGTCCCCGCGGCCCCCCTGGTCAACCAGGAACTAGAGGTTTCCCTGGATTTCCGGGTCCCATTGGGCTGGACGGCAAACCGGGCCACCCAGGACCAAAGGGCGACATGGGTCTGACGGGTCCCCCAGGACAGCCGGGACCCCAGGGACAAAAAGGAGAAAAGGGTCAGTGTGGAGAGTACCCACACCGGGAGTGCCTAAGCAGCATGCCAGCAGCTCTGCGCTCCAGCCAAATAATTGCCCTGAAGCTGCTGCCTCTCCTCAATTCAGTGCGACTGGCTCCACCCCCGGTCATAAAAAGGCGGACGTTCCAGGGCGAACAGAGCCAGGCCAGCATCCAAGGTCCACCAGGGCCCCCAGGCCCCCCTGGACCAAGTGGACCTCTGGGGCACCCAGGACTGCCAGGGCCTATGGGGCCACCTGGCTTACCTGGGCCTCCTGGACCAAAGGGAGACCCAGGGATCCAGGGCTACCACGGCCGGAAGGGAGAACGGGGCATGCCAGGGATGCCAGGCAAGCATGGAGCCAAGGGGGCGCCCGGAATTGCCGTGGCTGGGATGAAGGGTGAGCCAGGGATCCCAGGAACCAAGGGTGAGAAGGGGGCTGAAGGCTCCCCTGGGCTTCCTGGCCTCCTGGGGCAGAAGGGAGAGAAAGGCGATGCTGGCAACTCCATTGGAGGAGGCAGAGGGGAACCTGGCCCTCCAGGGCTCCCTGGGCCCCCAGGGCCAAAGGGAGAAGCAGGTGTCGATGGCCAGGTTGGCCCCCCAGGGCAGCCAGGAGACAAGGGGGAGCGTGGAGCAGCTGGAGAACAGGGACCAGATGGCCCCAAGGGCTCCAAGGGAGAACCAGGGAAAGGAGAGATGGTGGATTACAATGGAAACATCAATGAGGCTCTCCAGGAGATCCGGACGCTGGCCTTGATGGGGCCTCCTGGTCTTCCTGGGCAAATTGGCCCACCTGGAGCTCCAGGGATTCCAGGCCAGAAGGGGGAGATTGGACTGCCAGGCCCTCCAGGACACGATGGGGAAAAGGGACCTCGCGGTAAACCAGGAGACATGGGCCCTCCTGGTCCCCAAGGCCCCCCAGGAAAGGATGGACCTCCAGGAGTGAAGGGAGAAAACGGGCACCCAGGGAGCCCAGGAGAGAAGGGGGAAAAAGGGGAGACAGGACAAGCAGGCTCACCGGGAGAGAAAGGAGAAGCCGGGGAGAAGGGCAATCCAGGAGCAGAGGTTCCTGGGCTGCCAGGGCCAGAGGGGCCTCCCGGACCTCCGGGGCTCCAAGGTGTTCCTGGACCAAAGGGGGAAGCAGGACTAGATGGAGCAAAAGGAGAGAAAGGCTTCCAGGGAGAAAAAGGAGACCGTGGTCCCCTGGGACTACCCGGAGCTTCAGGTTTGGACGGCAGGCCTGGGCCACCGGGTACTCCAGGACCAATTGGAGTTCCAGGCCCAGCGGGACCAAAGGGCGAGAGGGGCAGCAAAGGAGACCCTGGGATGACAGGACCAACGGGAGCAGCTGGGCTTCCTGGTTTACATGGACCACCCGGGGACAAGGGAAACCGGGGGGAGAGGGGGAAGAAAGGCTCTAGAGGGCCTAAAGGGGATAAGGGAGACCAAGGAGCGCCTGGATTAGATGCCCCCTGCCCATTGGGCGAAGATGGCTTACCAGTCCAAGGCTGCTGGAACAAGTGATGCCTCTAACCTTGGATTGGCCTGTGTGTGTGTTTGTACATAGAATATTTATTTTTATACAGTTTTCACTTTTTGAAAATGCCAGAAGTATGATGCATCTTACAGATTATTAAAAAAGAAAGAAAAACCTGCATATTTTGTACAGAAAATATCAACCTCTTCCCTTTTGTTTACAAGATGTTTTGTATAAGCCTATGTCTCTAATACATTTTTTGTTTGGTCGTAATGTCTGCATGATATTTGTGCACATTTATTAAGTATCGAAGCTTAATAAATTATTGTGTCCTGGTGCCAAAGGGGGCCAGCCAGAACTGAGGTGCTGGCTAGCTCATGTGTGAATTCACATAAATGTAGAGGTCCATGATATTTGCTAAGCTAGGTGTGTCTAAGAGTATTTTAAACCCTTATGGATTTTCATTATTAAAGGAAATGAAACATGGCAA\n",
      "\n",
      "> 110011100,Gene_ID:NCBI:6867,Refseq_ID:NM_001122824,Species:Homo\n",
      "GAGTCCGCGAGCCGGGAGCGGGAGCAGCAGAGGTCTAGCAGCCGGGCGCCGCGGGCCGGGGGCCTGAGGAGGCCACAGGACGGGCGTCTTCCCGGCTAGTGGAGCCCGGCGCGGGGCCCGCTGCGGCCGCACCGTGAGGGGAGGAGGCCGAGGAGGACGCAGCGCCGGCTGCCGGCGGGAGGAAGCGCTCCACCAGGGCCCCCGACGGCACTCGTTTAACCACATCCGCGCCTCTGCTGGAAACGCTTGCTGGCGCCTGTCACCGGTTCCCTCCATTTTGAAAGGGAAAAAGGCTCTCCCCACCCATTCCCCTGCCCCTAGGAGCTGGAGCCGGAGGAGCCGCGCTCATGGCGTTCAGCCCGTGGCAGATCCTGTCCCCCGTGCAGTGGGCGAAATGGACGTGGTCTGCGGTACGCGGCGGGGCCGCCGGCGAGGACGAGGCTGGCGGGCCCGAGGGCGACCCCGAGGAGGAGGATTCGCAAGCCGAGACCAAATCCTTGAGTTTCAGGAGTGGCTGTAAGGTGAAGAAGCATGAAACTCAGTCTCTCGCCCTGGATGCATGTTCTCGGGATGAAGGGGCAGTGATCTCCCAGATTTCAGACATTTCTAATAGGGATGGCCATGCTACTGATGAGGAGAAACTGGCATCCACGTCATGTGGTCAGAAATCAGCTGGTGCCGAGGTGAAAGGTGAGCCAGAGGAAGACCTGGAGTACTTTGAATGTTCCAATGTTCCTGTGTCTACCATAAATCATGCGTTTTCATCCTCAGAAGCAGGCATAGAGAAGGAGACGTGCCAGAAGATGGAAGAAGACGGGTCCACTGTGCTTGGGCTGCTGGAGTCCTCTGCAGAGAAGGCCCCTGTGTCGGTGTCCTGTGGAGGTGAGAGCCCCCTGGATGGGATCTGCCTCAGCGAATCAGACAAGACAGCCGTGCTCACCTTAATAAGAGAAGAGATAATTACTAAAGAGATTGAAGCAAATGAATGGAAGAAGAAATACGAAGAGACCCGGCAAGAAGTTTTGGAGATGAGGAAAATTGTAGCTGAATATGAAAAGACTATTGCTCAAATGATTGAAGATGAACAAAGGACAAGTATGACCTCTCAGAAGAGCTTCCAGCAACTGACCATGGAGAAGGAACAGGCCCTGGCTGACCTTAACTCTGTGGAAAGGTCCCTTTCTGATCTCTTCAGGAGATATGAGAACCTGAAAGGTGTTCTGGAAGGGTTCAAGAAGAATGAAGAAGCCTTGAAGAAATGTGCTCAGGATTACTTAGCCAGAGTTAAACAAGAGGAGCAGCGATACCAGGCCCTGAAAATCCACGCAGAAGAGAAACTGGACAAAGCCAATGAAGAGATTGCTCAGGTTCGAACAAAAGCAAAGGCTGAGAGTGCAGCTCTCCATGCTGGACTCCGCAAAGAGCAGATGAAGGTGGAGTCCCTGGAAAGGGCCCTGCAGCAGAAGAACCAAGAAATTGAAGAACTGACAAAAATCTGTGATGAGCTGATTGCAAAGCTGGGAAAGACTGACTGAGACACTCCCCCTGTTAGCTCAACAGATCTGCATTTGGCTGCTTCTCTTGTGACCACAATTATCTTGCCTTATCCAGGAATAATTGCCCCTTTGCAGAGAAAAAAAAAAACTTAAAAAAAGCACATGCCTACTGCTGCCTGTCCCGCTTTGCTGCCAATGCAACAGCCCTGGAAGAAACCCTAGAGGGTTGCATAGTCTAGAAAGGAGTGTGACCTGACAGTGCTGGAGCCTCCTAGTTTCCCCCTATGAAGGTTCCCTTAGGCTGCTGAGTTTGGGTTTGTGATTTATCTTTAGTTTGTTTTAAAGTCATCTTTACTTTCCCAAATGTGTTAAATTTGTAACTCCTCTTTGGGGTCTTCTCCACCACCTGTCTGATTTTTTTGTGATCTGTTTAATCTTTTAATTTTTTAGTATCAGTGGTTTTATTTAAGGAGACAGTTTGGCCTATTGTTACTTCCAATTTATAATCAAGAAGGGGCTCTGGATCCCCTTTTAAATTACACACACTCTCACACACATACATGTATGTTTATAGATGCTGCTGCTCTTTTCCCTGAAGCATAGTCAAGTAAGAACTGCTCTACAGAAGGACATATTTCCTTGGATGTGAGACCCTATTTTGAAATAGAGTCCTGACTCAGAACACCAACTTAAGAATTTGGGGGATTAAAGATGTGAAGACCACAGTCTTGGGTTTTCATATCTGGAGAAGACTATTTGCCATGACGTTTTGTTGCCCTGGTATTTGGACACTCCTCAGCTTTAATGGGTGTGGCCCCTTTAGGGTTAGTCCTCAGACTAATGATAGTGTCTGCTTTCTGCATGAACGGCAATATGGGACTCCCTCCAAGCTAGGGTTTGGCAAGTCTGCCCTAGAGTCATTTACTCTCCTCTGCCTCCATTTGTTAATACAGAATCAACATTTAGTCTTCATTATCTTTTTTTTTTTTTTTGAGACAGAGTTTCGATCTATTTTAAGTATGTGAAGAAAATCTACTTGTAAAAGGCTCAGATCTTAATTAAAAGGTAATTGTAGCACATTACCAATTATAAGGTGAAGAAATGTTTTTTTCCCAAGTGTGATGCATTGTTCTTCAGATGTTGAAAAGAAAGCAAAAAATACCTTCTAACTTAAGACAGAATTTTTAACAAAATGAGCAGTAAAAGTCACATGAACCACTCCAAAAATCAGTGCATTTTGCATATTTTTAAACAAAGACAGCTTGTTGAATACTGAGAAGAGGAGTGCAAGGAGAAGGTCTGTACTAACAAAGCCAAATTCCTCAAGCTCTTACTGGACTCAGTTCAGAGTGGTGGGCCATTAACCCCAACATGGAATTTTTCCATATAAATCTCAATGAATTCCCTTTCATTTGAATAGGCAAACCCAAATCCATGCAAGTGTTTTAAAGCACTGTCCTGTCTTAATCTTACATGCTGAAAGTCTTCATGGTGATATGCACTATATTCAGTATACGTATGTTTTCCTACTTCTCTTGTAAAACTGTTGCATGATCCAACTTCAGCAATGAATTGTGCCTAGTGGAGAACCTCTATAGATCTTAAAAAATGAATTATTCTTTAGCAGTGTATTACTCACATGGGTGCAATCTTTAGCCCCAGGGAGGTCAATAATGTCTTTTAAAGCCAGAAGTCACATTTTACCAATATGCATTTATCATAATTGGTGCTTAGGCTGTATATTCAAGCCTGTTGTCTTAACATTTTGTATAAAAAAGAACAACAGAAATTATCTGTCATTTGAGAAGTGGCTTGACAATCATTTGAGCTTTGAAAGCAGTCACTGTGGTGTAATATGAATGCTGTCCTAGTGGTCATAGTACCAAGGGCACGTGTCTCCCCTTGGTATAACTGATTTCCTTTTTAGTCCTCTACTGCTAAATAAGTTAATTTTGCATTTTGCAGAAAGAAACATTGATTGCTAAATCTTTTTGCTGCTGTGTTTTGGTGTTTTCATGTTTACTTGTTTTATATTGATCTGTTTTAAGTATGAGAGGCTTATAGTGCCCTCCATTGTAAATCCATAGTCATCTTTTTAAGCTTATTGTGTTTAAGAAAGTAGCTATGTGTTAAACAGAGGTGATGGCAGCCCTTCCCTAGCACACTGGTGGAAGAGACCCCTTAAGAACCTGACCCCAGTGAATGAAGCTGATGCACAGGGAGCACCAAAGGACCTTCGTTAAGTGATAATTGTCCTGGCCTCTCAGCCATGACCGTTATGAGGAAATATCCCCCATTCGAACTTAACAGATGCCTCCTCTCCAAAGAGAATTAAAATCGTAGCTTGTACAGATCAAGAGAATATACTGGGCAGAATGAAGTATGTTTGTTTATTTTTCTTTAAAAATAAAGGATTTTGGAACTCTGGAGAGTAAGAATATAGTATAGAGTTTGCCTCAACACATGTGAGGGCCAAATAACCTGCTAGCTAGGCAGTAATAAACTCTGTTACAGAAGAGAAAAAGGGCCGGGCACAGTGGCTTATTCCTGTAATCCCAACACTGTGGAAGGCCGAGGCAGGAGGATCACTTGAGTCCAGGAGTTTGAAACCTACCTAGGCAACATGGTGAAACCTTGTCTCTACCAAAATAAAAATTAGCTGGGCATGGTGGCACGTGCCTGTGGTCCCAGCTACTTGGGAGGCTGAGGTGGGAGCCTGGGAGGTCAAGGCTGCAGTGAGCCATGATCATGCCACTGCACTCCATCCTGGGTGACAGCAAGATCTTGTCTCAAAAAAAAAAAAAAAAAAAAAAAAACCAGGAGTGAAAAAGGAAAGTAGAAGGCAGCTGCTGGCCTAGATGTTGGTTTGGGAATATTAGGTGATCCTGTTGAGATTCTGGATCCAGAGCAATTTCTTTAGCTTTTGACTTTGCCAAAGTGTAGATAGCCTTTATCCAGCAGTATTTTAAGTGGGGAATGCAACGTGAGGCCAACTGAACAATTCCCCCCGTGGCTGCCCAGATAGTCACAGTCAAGGTTGGAGAGTCTCCTTCCAGCCAGTGACCTACCCAAACCTTTTGTTCTGTAAAACTGCTCTGGAAATACCGGGAAGCCCAGTTTTCTCACGTGGTTTCTAGCTTCTTCAGACTCAGCCCAAATTAGGAAGTGCAGAAGCACATGATGGTGAAAAACCTAGGATTTGGCAGCCTTCCAGAATGGTATGGAATCTGAGGGAAGATTTATGTTTCGTTTTGGAGGATAGCTCAAGTTGAATTTTCTTTCCAGCCAGTTACCCTTTCAACCTACCCATACTTTGTACAACTCTTACACAAATACTTAGATATTTATTAGATAGCCCTGAATTCACTCTAATTATAAACAGGGAGTGTAAACTGCCCCCAGATGTTCCTGGGCTGGGTAAAAGCAGCTGGAGTGAAGCACTCATTTTCCATAAAGGTAACAAAGGGCAGCTCAGTGGTTACTCAAGCTCAAAAGGGTTTTTTTAAGAGCAAGCATTGGTTAAGTCTGTGTATACTGAGTTGGAAGTGATTTCAGCACATTCTTTTTTAGTGGAGTGAAAGTTCTGAAGCCCCCTTTTAACTTCCTCTTGGTTTTTCATTATAATTGGTAGCCATCTCATGAACTGTCTCTGACTGTTGTCTCTTTGTGGTCATGTGATTGTGAGCTTGCTTTCTGACTTGCATTTCTGACTTTATCCTGTTGTTAGGAAGATAGAAACTAGGTTTTGAAAGATTACATGATTCAAGCGAGGGATTTTAAAGTAAAGATGTATTTATTCTGAAGAATCTAAAAGATAACAGATTATTTGCTTATGAAAGAACAATATAGTCTGGGAATCCCAGAATGTCAAGCCAAAGGTCTAAGAAGTCATCTCCTTCAAATACTTTAATAAAGAAGTATTTCGAGGAGATATCTGTCCAAAAAGGTTTGACTGGCCTCCAGATTCCAGTTATTTTTAAAAAGCAACTTACCACTAAATCCTTGAGTCTCCATAGAGTAACAGTAAAGAAACTGATGTAACAGACTCTCCTCTCAAAGGATCTCCTCTGGAAGAGACTATCAGCGGCAGCATTCTCCAGGGAAGACCCATCCCCTAGTGCCAGAGCTTGCATCCTGGAGACTAAAGATTGCACTTTTTTGTAGTTTTTTGTCCAAATGCAATCCCATTTCTGTGCCTCTTAGCATGCAGTTAGATTTGGACAAACAAGATTCCTAAGGAATGACTTTATTAACTATAATATGGTTACAGCTATTATATAAATATATATTCTGGTTATAGTTCTAATATGGAGATGTTGTGTGCAATGCTGGCCTGTGGTGGTCTGTGTAATGCTTTAACTTGTATGGAGGAGGCCAGGCTCAGAGCTGAGATGTGGCCTGAACCTTCCCTGTATCGATCCTTTAATTTAGAACTGTCAAGATGTCACTTTCTCCCCCTCTGCCTTTTAGTGGTATCTGACATATACTCAAAACAGTAATTTCCTGGTCACATCATTAACTGCTAATTCTGTATTTATAAAGAATTTTCAGATGGACATGTACAAATTTGAACTCAAACCATCCCCAGTCCAGATACAGGGCAGCGTGTAGGTGACCACACCAGAGCCTCAGCCTCGGTCCTTCTCAGCCGTCGGGATAGGATCCAGGCATTTCTTTTAAATCTCAGAGGTAGCAGTAAACTTTTCAGTATTGCTGTTAGCAAGTGTGTGTTTGCCAATAGATACCCATTATACTAATGTGCCAAGTAAATGTTCATTGCACATCTGCTTCCACTGTGTTCCCACGGGTGCCATGAAGTGTGTGAGGAGCCCCTCATCTGGAGGGATGAGTGCTGCGTTGACTACTGCTATCAGGATTGTGTTGTGTGGAATATTCATCTACATAAATTTTATATGCACAGTAATTTCCCTTTTTATATGTCAAGTAACTATTTGTAAAAGTTATACTCACAAATTATTATAATGATTACTAATATATTTTTTCCATGTTTCATTGCCTGAATAAAAACTGTTTACCACTGTTA\n",
      "\n",
      "> 110101000,Gene_ID:NCBI:6597,Refseq_ID:NM_001128844,Species:Homo\n",
      "GGGCGCGCGCGCGAGGCTTCCCCTCGTTTGGCGGCGGCGGCGGCTTCTTTGTTTCGTGAAGAGAAGCGAGACGCCCATTCTGCCCCCGGCCCCGCGCGGAGGGGCGGGGGAGGCGCCGGGAAGTCGACGGCGCCGGCGGCTCCTGCGTCTCGCCCTTTTGCCCAGGCTAGAGTGCAGTGGTGCGGTCATGGTTCACTGCAGCCTCAACCTCCTGGACTCAGCAGGAGGCCACTGTCTGCAGCTCCCGTGAAGATGTCCACTCCAGACCCACCCCTGGGCGGAACTCCTCGGCCAGGTCCTTCCCCGGGCCCTGGCCCTTCCCCTGGAGCCATGCTGGGCCCTAGCCCGGGTCCCTCGCCGGGCTCCGCCCACAGCATGATGGGGCCCAGCCCAGGGCCGCCCTCAGCAGGACACCCCATCCCCACCCAGGGGCCTGGAGGGTACCCTCAGGACAACATGCACCAGATGCACAAGCCCATGGAGTCCATGCATGAGAAGGGCATGTCGGACGACCCGCGCTACAACCAGATGAAAGGAATGGGGATGCGGTCAGGGGGCCATGCTGGGATGGGGCCCCCGCCCAGCCCCATGGACCAGCACTCCCAAGGTTACCCCTCGCCCCTGGGTGGCTCTGAGCATGCCTCTAGTCCAGTTCCAGCCAGTGGCCCGTCTTCGGGGCCCCAGATGTCTTCCGGGCCAGGAGGTGCCCCGCTGGATGGTGCTGACCCCCAGGCCTTGGGGCAGCAGAACCGGGGCCCAACCCCATTTAACCAGAACCAGCTGCACCAGCTCAGAGCTCAGATCATGGCCTACAAGATGCTGGCCAGGGGGCAGCCCCTCCCCGACCACCTGCAGATGGCGGTGCAGGGCAAGCGGCCGATGCCCGGGATGCAGCAGCAGATGCCAACGCTACCTCCACCCTCGGTGTCCGCAACAGGACCCGGCCCTGGCCCTGGCCCTGGCCCCGGCCCGGGTCCCGGCCCGGCACCTCCAAATTACAGCAGGCCTCATGGTATGGGAGGGCCCAACATGCCTCCCCCAGGACCCTCGGGCGTGCCCCCCGGGATGCCAGGCCAGCCTCCTGGAGGGCCTCCCAAGCCCTGGCCTGAAGGACCCATGGCGAATGCTGCTGCCCCCACGAGCACCCCTCAGAAGCTGATTCCCCCGCAGCCAACGGGCCGCCCTTCCCCCGCGCCCCCTGCCGTCCCACCCGCCGCCTCGCCCGTGATGCCACCGCAGACCCAGTCCCCCGGGCAGCCGGCCCAGCCCGCGCCCATGGTGCCACTGCACCAGAAGCAGAGCCGCATCACCCCCATCCAGAAGCCGCGGGGCCTCGACCCTGTGGAGATCCTGCAGGAGCGCGAGTACAGGCTGCAGGCTCGCATCGCACACCGAATTCAGGAACTTGAAAACCTTCCCGGGTCCCTGGCCGGGGATTTGCGAACCAAAGCGACCATTGAGCTCAAGGCCCTCAGGCTGCTGAACTTCCAGAGGCAGCTGCGCCAGGAGGTGGTGGTGTGCATGCGGAGGGACACAGCGCTGGAGACAGCCCTCAATGCTAAGGCCTACAAGCGCAGCAAGCGCCAGTCCCTGCGCGAGGCCCGCATCACTGAGAAGCTGGAGAAGCAGCAGAAGATCGAGCAGGAGCGCAAGCGCCGGCAGAAGCACCAGGAATACCTCAATAGCATTCTCCAGCATGCCAAGGATTTCAAGGAATATCACAGATCCGTCACAGGCAAAATCCAGAAGCTGACCAAGGCAGTGGCCACGTACCATGCCAACACGGAGCGGGAGCAGAAGAAAGAGAACGAGCGGATCGAGAAGGAGCGCATGCGGAGGCTCATGGCTGAAGATGAGGAGGGGTACCGCAAGCTCATCGACCAGAAGAAGGACAAGCGCCTGGCCTACCTCTTGCAGCAGACAGACGAGTACGTGGCTAACCTCACGGAGCTGGTGCGGCAGCACAAGGCTGCCCAGGTCGCCAAGGAGAAAAAGAAGAAAAAGAAAAAGAAGAAGGCAGAAAATGCAGAAGGACAGACGCCTGCCATTGGGCCGGATGGCGAGCCTCTGGACGAGACCAGCCAGATGAGCGACCTCCCGGTGAAGGTGATCCACGTGGAGAGTGGGAAGATCCTCACAGGCACAGATGCCCCCAAAGCCGGGCAGCTGGAGGCCTGGCTCGAGATGAACCCGGGGTATGAAGTAGCTCCGAGGTCTGATAGTGAAGAAAGTGGCTCAGAAGAAGAGGAAGAGGAGGAGGAGGAAGAGCAGCCGCAGGCAGCACAGCCTCCCACCCTGCCCGTGGAGGAGAAGAAGAAGATTCCAGATCCAGACAGCGATGACGTCTCTGAGGTGGACGCGCGGCACATCATTGAGAATGCCAAGCAAGATGTCGATGATGAATATGGCGTGTCCCAGGCCCTTGCACGTGGCCTGCAGTCCTACTATGCCGTGGCCCATGCTGTCACTGAGAGAGTGGACAAGCAGTCAGCGCTTATGGTCAATGGTGTCCTCAAACAGTACCAGATCAAAGGTTTGGAGTGGCTGGTGTCCCTGTACAACAACAACCTGAACGGCATCCTGGCCGACGAGATGGGCCTGGGGAAGACCATCCAGACCATCGCGCTCATCACGTACCTCATGGAGCACAAACGCATCAATGGGCCCTTCCTCATCATCGTGCCTCTCTCAACGCTGTCCAACTGGGCGTACGAGTTTGACAAGTGGGCCCCCTCCGTGGTGAAGGTGTCTTACAAGGGATCCCCAGCAGCAAGACGGGCCTTTGTCCCCCAGCTCCGGAGTGGGAAGTTCAACGTCTTGCTGACGACGTACGAGTACATCATCAAAGACAAGCACATCCTCGCCAAGATCCGTTGGAAGTACATGATTGTGGACGAAGGTCACCGCATGAAGAACCACCACTGCAAGCTGACGCAGGTGCTCAACACGCACTATGTGGCACCCCGCCGCCTGCTGCTGACGGGCACACCGCTGCAGAACAAGCTTCCCGAGCTCTGGGCGCTGCTCAACTTCCTGCTGCCCACCATCTTCAAGAGCTGCAGCACCTTCGAGCAGTGGTTTAACGCACCCTTTGCCATGACCGGGGAAAAGGTGGACCTGAATGAGGAGGAAACCATTCTCATCATCCGGCGTCTCCACAAAGTGCTGCGGCCCTTCTTGCTCCGACGACTCAAGAAGGAAGTCGAGGCCCAGTTGCCCGAAAAGGTGGAGTACGTCATCAAGTGCGACATGTCTGCGCTGCAGCGAGTGCTCTACCGCCACATGCAGGCCAAGGGCGTGCTGCTGACTGATGGCTCCGAGAAGGACAAGAAGGGCAAAGGCGGCACCAAGACCCTGATGAACACCATCATGCAGCTGCGGAAGATCTGCAACCACCCCTACATGTTCCAGCACATCGAGGAGTCCTTTTCCGAGCACTTGGGGTTCACTGGCGGCATTGTCCAAGGGCTGGACCTGTACCGAGCCTCGGGTAAATTTGAGCTTCTTGATAGAATTCTTCCCAAACTCCGAGCAACCAACCACAAAGTGCTGCTGTTCTGCCAAATGACCTCCCTCATGACCATCATGGAAGATTACTTTGCGTATCGCGGCTTTAAATACCTCAGGCTTGATGGAACCACGAAGGCGGAGGACCGGGGCATGCTGCTGAAAACCTTCAACGAGCCCGGCTCTGAGTACTTCATCTTCCTGCTCAGCACCCGGGCTGGGGGGCTCGGCCTGAACCTCCAGTCGGCAGACACTGTGATCATTTTTGACAGCGACTGGAATCCTCACCAGGACCTGCAAGCGCAGGACCGAGCCCACCGCATCGGGCAGCAGAACGAGGTGCGTGTGCTCCGCCTCTGCACCGTCAACAGCGTGGAGGAGAAGATCCTAGCTGCAGCCAAGTACAAGCTCAACGTGGACCAGAAGGTGATCCAGGCCGGCATGTTCGACCAGAAGTCCTCCAGCCATGAGCGGCGCGCCTTCCTGCAGGCCATCCTGGAGCACGAGGAGCAGGATGAGAGCAGACACTGCAGCACGGGCAGCGGCAGTGCCAGCTTCGCCCACACTGCCCCTCCGCCAGCGGGCGTCAACCCCGACTTGGAGGAGCCACCTCTAAAGGAGGAAGACGAGGTGCCCGACGACGAGACCGTCAACCAGATGATCGCCCGGCACGAGGAGGAGTTTGATCTGTTCATGCGCATGGACCTGGACCGCAGGCGCGAGGAGGCCCGCAACCCCAAGCGGAAGCCGCGCCTCATGGAGGAGGACGAGCTCCCCTCGTGGATCATCAAGGACGACGCGGAGGTGGAGCGGCTGACCTGTGAGGAGGAGGAGGAGAAGATGTTCGGCCGTGGCTCCCGCCACCGCAAGGAGGTGGACTACAGCGACTCACTGACGGAGAAGCAGTGGCTCAAGGCCATCGAGGAGGGCACGCTGGAGGAGATCGAAGAGGAGGTCCGGCAGAAGAAATCATCACGGAAGCGCAAGCGAGACAGCGACGCCGGCTCCTCCACCCCGACCACCAGCACCCGCAGCCGCGACAAGGACGACGAGAGCAAGAAGCAGAAGAAGCGCGGGCGGCCGCCTGCCGAGAAACTCTCCCCTAACCCACCCAACCTCACCAAGAAGATGAAGAAGATTGTGGATGCCGTGATCAAGTACAAGGACAGCAGCAGTGGACGTCAGCTCAGCGAGGTCTTCATCCAGCTGCCCTCGCGAAAGGAGCTGCCCGAGTACTACGAGCTCATCCGCAAGCCCGTGGACTTCAAGAAGATAAAGGAGCGCATTCGCAACCACAAGTACCGCAGCCTCAACGACCTAGAGAAGGACGTCATGCTCCTGTGCCAGAACGCACAGACCTTCAACCTGGAGGGCTCCCTGATCTATGAAGACTCCATCGTCTTGCAGTCGGTCTTCACCAGCGTGCGGCAGAAAATCGAGAAGGAGGATGACAGTGAAGGCGAGGAGAGTGAGGAGGAGGAAGAGGGCGAGGAGGAAGGCTCCGAATCCGAATCTCGGTCCGTCAAAGTGAAGATCAAGCTTGGCCGGAAGGAGAAGGCACAGGACCGGCTGAAGGGCGGCCGGCGGCGGCCGAGCCGAGGGTCCCGAGCCAAGCCGGTCGTGAGTGACGATGACAGTGAGGAGGAACAAGAGGAGGACCGCTCAGGAAGTGGCAGCGAAGAAGACTGAGCCCCGACATTCCAGTCTCGACCCCGAGCCCCTCGTTCCAGAGCTGAGATGGCATAGGCCTTAGCAGTAACGGGTAGCAGCAGATGTAGTTTCAGACTTGGAGTAAAACTGTATAAACAAAAGAATCTTCCATATTTATACAGCAGAGAAGCTGTAGGACTGTTTGTGACTGGCCCTGTCCTGGCATCAGTAGCATCTGTAACAGCATTAACTGTCTTAAAGAGAGAGAGAGAGAATTCCGAATTGGGGAACACACGATACCTGTTTTTCTTTTCCGTTGCTGGCAGTACTGTTGCGCCGCAGTTTGGAGTCACTGTAGTTAAGTGTGGATGCATGTGCGTCACCGTCCACTCCTCCTACTGTATTTTATTGGACAGGTCAGACTCGCCGGGGGCCCGGCGAGGGTATGTCAGTGTCACTGGATGTCAAACAGTAATAAATTAAACCAACAACAAAA\n",
      "\n",
      "> 111101000,Gene_ID:NCBI:55160,Refseq_ID:NM_001011722,Species:Homo\n",
      "GTGTGTAGCTGGGACGGTGCTGGTCTGAGCTGGACCTTGTCTGATGGCTTCCTCCAACCCTCCTCCACAGCCTGCCATAGGAGATCAGCTGGTTCCAGGAGTCCCAGGCCCCTCCTCTGAGGCAGAGGACGACCCAGGAGAGGCGTTTGAGTTTGATGACAGTGATGATGAAGAGGACACCAGCGCAGCCCTGGGCGTCCCCAGCCTTGCTCCTGAGAGGGACACAGACCCCCCACTGATCCACTTGGACTCCATCCCTGTCACTGACCCAGACCCAGCAGCTGCTCCACCCGGCACAGGGGTGCCAGCCTGGGTGAGCAATGGGGATGCAGCGGACGCAGCCTTCTCCGGGGCCCGGCACTCCAGCTGGAAGCGGAAGAGTTCCCGTCGCATTGACCGGTTCACTTTCCCCGCCCTGGAAGAGGATGTGATTTATGACGACGTCCCCTGCGAGAGCCCAGATGCGCATCAGCCCGGGGCAGAGAGGAACCTGCTCTACGAGGATGCGCACCGGGCTGGGGCCCCTCGGCAGGCGGAGGACCTAGGCTGGAGCTCCAGTGAGTTCGAGAGCTACAGCGAGGACTCGGGGGAGGAGGCCAAGCCGGAGGTCGAGGTCGAGCCCGCCAAGCACCGAGTGTCCTTCCAGCCCAAGATGACCCAGCTCATGAAGGCCGCCAAGAGCGGGACCAAGGATGGGCTGGAGAAGACACGGATGGCCGTGATGCGCAAAGTCTCCTTCCTGCACAGGAAGGACGTCCTCGGTGACTCGGAGGAGGAGGACATGGGGCTCCTGGAGGTCAGCGTTTCGGACATCAAGCCCCCAGCCCCAGAGCTGGGCCCCATGCCAGAGGGCCTGAGCCCTCAGCAGGTGGTCCGGAGGCATATCCTGGGCTCCATCGTGCAGAGCGAAGGCAGCTACGTGGAGTCTCTGAAGCGGATACTCCAGGACTACCGCAACCCCCTGATGGAGATGGAGCCCAAGGCGCTGAGCGCCCGCAAGTGCCAGGTGGTGTTCTTCCGCGTGAAGGAGATCCTGCACTGCCACTCCATGTTCCAGATCGCCCTGTCCTCCCGCGTGGCTGAGTGGGATTCCACCGAGAAGATCGGGGACCTCTTCGTGGCCTCGTTTTCCAAGTCCATGGTGCTAGATGTGTACAGTGACTACGTGAACAACTTCACCAGTGCCATGTCCATCATCAAGAAGGCCTGCCTCACCAAGCCTGCCTTCCTCGAGTTCCTCAAGCGACGGCAGGTGTGCAGCCCAGACCGTGTCACCCTCTACGGGCTGATGGTCAAGCCCATCCAGAGGTTCCCACAGTTCATACTCCTGCTTCAGGACATGCTGAAGAACACCCCCAGGGGCCATCCGGACAGGCTGTCGCTGCAGCTGGCCCTCACAGAGCTGGAGACGCTGGCTGAGAAGCTGAACGAGCAGAAGCGGCTGGCTGACCAGGTGGCTGAGATCCAGCAGCTGACCAAGAGCGTCAGTGACCGCAGCAGCCTCAACAAGCTGTTGACCTCAGGCCAGCGGCAGCTGCTCCTGTGTGAGACGTTGACGGAGACCGTGTACGGTGACCGCGGGCAGCTAATTAAGTCCAAGGAGCGTCGGGTCTTCCTGCTCAACGACATGCTTGTCTGTGCCAACATCAACTTCAAGCCTGCCAACCACAGGGGCCAGCTGGAGATCAGCAGCCTGGTGCCCCTGGGGCCCAAGTATGTGGTGAAGTGGAACACGGCGCTGCCCCAGGTGCAGGTGGTGGAGGTGGGCCAGGACGGTGGCACCTATGACAAGGACAATGTGCTCATCCAGCACTCAGGCGCCAAGAAGGCCTCTGCCTCAGGGCAGGCTCAGAATAAGGTGTACCTCGGCCCCCCACGCCTCTTCCAGGAGCTGCAGGACCTGCAGAAGGACCTGGCCGTGGTGGAGCAGATCACGCTTCTCATCAGCACGCTGCACGGCACCTACCAGAACCTGAACATGACTGTGGCTCAAGACTGGTGCCTGGCCCTGCAGAGGCTGATGCGGGTGAAGGAGGAAGAGATCCACTCGGCCAACAAGTGCCGTCTCAGGCTCCTGCTTCCTGGGAAACCCGACAAGTCCGGCCGCCCCATTAGCTTCATGGTGGTTTTCATCACCCCCAACCCCCTGAGCAAGATTTCCTGGGTCAACAGGTTACATTTGGCCAAAATCGGACTCCGGGAGGAGAACCAGCCAGGCTGGCTATGCCCGGATGAGGACAAGAAGAGCAAAGCCCCATTCTGGTGCCCGATCCTGGCCTGCTGCATCCCTGCCTTCTCCTCCCGGGCACTCAGCCTGCAGCTTGGGGCCCTGGTCCACAGTCCTGTCAACTGTCCCCTGCTGGGTTTCTCAGCAGTCAGCACCTCCCTTCCACAGGGCTACCTCTGGGTCGGGGGCGGACAGGAAGGCGCAGGGGGCCAGGTGGAAATCTTTTCCTTGAACCGGCCCTCGCCCCGCACCGTCAAGTCCTTCCCACTGGCAGCCCCTGTGCTCTGCATGGAGTATATCCCGGAGCTGGAGGAGGAGGCGGAGAGCAGAGACGAGAGCCCGACAGTTGCTGACCCCTCGGCCACGGTGCATCCAACCATCTGCCTCGGGCTCCAGGATGGCAGCATCCTCCTCTACAGCAGTGTGGACACTGGCACCCAGTGCCTGGTGAGCTGCAGGAGCCCAGGTCTGCAGCCTGTGCTCTGCCTGCGACACAGCCCCTTCCACCTGCTCGCTGGCCTGCAGGATGGGACCCTTGCTGCTTACCCTCGGACCAGCGGAGGTGTCCTGTGGGACCTGGAGAGCCCTCCCGTGTGCCTGACTGTGGGGCCCGGGCCTGTCCGCACCCTGTTGAGCCTGGAGGATGCCGTGTGGGCCAGCTGTGGGCCCCGGGTCACTGTCCTGGAAGCCACCACCCTGCAGCCTCAGCAAAGCTTCGAGGCGCACCAGGACGAGGCAGTGAGCGTGACACACATGGTGAAGGCGGGCAGCGGCGTCTGGATGGCCTTCTCCTCCGGCACCTCCATCCGCCTCTTCCACACTGAGACCCTGGAGCATCTGCAAGAGATCAACATCGCCACCAGGACCACCTTCCTCCTGCCAGGCCAGAAGCACTTGTGTGTCACCAGCCTCCTGATCTGCCAGGGTCTGCTCTGGGTGGGCACTGACCAGGGTGTCATCGTCCTGCTGCCCGTGCCTCGGCTGGAAGGCATCCCCAAGATCACAGGGAAAGGCATGGTCTCACTCAACGGGCACTGTGGGCCTGTGGCCTTCCTGGCTGTGGCTACCAGCATCCTGGCCCCTGACATCCTGCGGAGTGACCAGGAGGAGGCTGAGGGGCCCCGGGCTGAGGAGGACAAGCCAGACGGGCAGGCACACGAGCCCATGCCCGATAGCCACGTGGGCCGAGAGCTGACCCGCAAGAAGGGCATCCTCTTGCAGTACCGCCTGCGCTCCACCGCACACCTCCCGGGCCCGCTGCTCTCCATGCGGGAGCCGGCGCCTGCTGATGGCGCAGCTTTGGAGCACAGCGAGGAGGACGGCTCCATTTACGAGATGGCCGACGACCCCGACATCTGGGTGCGCAGCCGGCCCTGCGCCCGCGACGCCCACCGCAAGGAGATTTGCTCTGTGGCCATCATCTCCGGCGGGCAGGGCTACCGCAACTTTGGCAGCGCTCTGGGCAGCAGTGGGAGGCAGGCCCCGTGTGGGGAGACGGACAGCACCCTCCTCATCTGGCAGGTGCCCTTGATGCTATAGCGCCTCCCCTCTCCCCTCAGAGGGCACAGCTGCAGGCCTGACCAAGGCCACGCCCGGCTCTCGTGCTCTAGGACCTGCACGGGACTTGTGGATGGGCCTGGACTCTCCAGAAACTACTTGGGCAGAGCAAAGGAAAACCTCTTGTTTTAAAAAAATTTTTTTCAGAGTGTTTTGGGGAGGAGTTTTAGGGCTTGGGGAGAGGGAGGACACATCTGGAGGAAATGGCCTTCTTTTTAAAAGCAAAAAACACAAAACCTCACAACTGCCTGGCAAGCCCAGTATCACTTGTTTGGGCCCTAGCGGGACTCCAAGGCAGCCACACGCCCCTCCTGGAAGGGTGTGTGCGTGTGAGTGTGTGCGAGTGTGTGGGCTGGTGTGTGAATATCTATAAATAAGTATATATGGTGTATATTATATGTGTATAAATAAAGTCTGTACATATTGGAGCTCTGGGAGATGCTGGAATAAAAGACAAGAGTTACATCTGGACTTGGAAAAAAAAA\n",
      "\n",
      "> 110100000,Gene_ID:NCBI:728642,Refseq_ID:NM_001313896,Species:Homo\n",
      "ATATCATCATTAGGCGTCAACACAGGAAGTGAGGATACTTCTGGCGAGCGCCGGTTGCTGTTTCTTCTCAGGCTCAGGGACCGGCCGCGGCCCCGTAGGGTGTTTTAACTCAAATGGGTGATGAAAAGGACTCTTGGAAAGTGAAAACTTTAGATGAAATTCTTCAGGAAAAGAAACGAAGGAAGGAACAAGAGGAGAAAGCAGAGATAAAACGCTTAAAAAATTCTGATGACCGGGATTCCAAGCGGGATTCCCTTGAGGAGGGGGAGCTGAGAGATCACTGCATGGAGATCACAATAAGGAACTCCCCGTATAGAAGAGAAGACTCAATGGAAGACAGAGGAGAAGAAGATGATTCTTTGGCCATCAAACCACCCCAGCAAATGTCTCGGAAAGAAAAAGTTCATCACAGAAAAGATGAAAAGAGAAAAGAAAAATGGAAGCATGCTAGAGTGAAAGAAAGAGAGCACGAACGTCGGAAACGACATCGAGAAGAACAGGATAAAGCTCGCCGGGAATGGGAAAGACAGAAGAGAAGGGAAATGGCAAGGGAGCATTCCAGGAGAGAAAGGGACCGCTTGGAGCAGTTAGAAAGGAAGCGGGAGCGGGAGCGCAAGATGCGGGAGCAGCAGAAGGAGCAGCGGGAGCAGAAGGAGCGCGAGCGGCGGGCGGAGGAGCGGCGCAAGGAGCGGGAGGCCCGCAGGGAAGTGTCTGCACATCACCGAACGATGAGAGAGGACTACAGCGACAAAGTGAAAGCCAGCCACTGGAGTCGCAGCCCGCCTCGGCCGCCGCGGGAGCGGTTCGAGTTGGGAGACGGCCGGAAGCCAGGTGAGGCCAGGCCGGCGCCTGCGCAGAAGCCAGCACAGTTAAAAGAAGAGAAAATGGAAGAAAGGGACCTGCTGTCCGACTTACAGGACATCAGCGACAGCGAGAGGAAGACCAGCTCGGCCGAGTCCTCGTCAGCGGAATCAGGCTCAGGTTCTGAGGAAGAAGAGGAGGAGGAGGAAGAGGAGGAGGAGGAAGGGAGCACCAGTGAAGAATCAGAGGAGGAGGAGGAGGAAGAGGAAGAGGAGGAGGAGGAGACCGGCAGCAACTCTGAGGAGGCATCAGAGCAGTCTGCCGAAGAAGTAAGTGAGGAAGAAATGAGTGAAGATGAAGAACGAGAAAATGAAAACCACCTCTTGGTTGTTCCAGAGTCACGGTTCGACCGAGATTCCGGGGAGAGTGAAGAAGCAGAGGAAGAAGTGGGTGAGGGAACGCCGCAGAGCAGCGCCCTGACAGAGGGCGACTATGTGCCCGACTCCCCTGCCCTGTTGCCCATCGAGCTCAAGCAGGAGCTGCCCAAGTACCTGCCGGCCCTGCAGGGCTGCCGGAGCGTCGAGGAGTTCCAGTGCCTGAACAGGATCGAGGAGGGCACCTATGGAGTGGTCTACAGAGCAAAAGACAAGAAAACAGATGAAATTGTGGCTCTAAAGCGGCTGAAGATGGAGAAGGAGAAGGAGGGCTTCCCGATCACGTCCCTGAGGGAGATCAACACCATCCTCAAGGCCCAGCATCCCAACATTGTCACCGTTAGAGAGATTGTGGTGGGCAGCAACATGGACAAGATCTACATCGTGATGAACTACGTGGAGCACGACCTCAAGAGCCTGATGGAGACCATGAAACAGCCCTTCCTGCCAGGGGAGGTGAAGACCCTGATGATCCAGCTGCTGCGGGGGGTGAAACACCTGCACGACAACTGGATCCTGCACCGTGACCTCAAGACGTCCAACCTGCTGCTGAGCCACGCCGGCATCCTCAAGGTGGGTGATTTTGGGCTGGCGCGGGAGTACGGATCCCCTCTGAAGGCCTACACCCCGGTCGTGGTGACCCAGTGGTACCGCGCCCCAGAGCTGCTGCTTGGTGCCAAGGAATACTCCACGGCCGTGGACATGTGGTCAGTGGGCTGCATCTTCGGGGAGCTGCTGACTCAGAAGCCTCTGTTCCCCGGGAATTCGGAAATCGATCAGATCAACAAAGTGTTCAAGGAGCTGGGGACCCCCAGTGAGAAAATCTGGCCCGGCTACAGTGAGCTCCCAGTAGTCAAAAAGATGACCTTCAGCGAGCACCCCTACAACAACCTCCGCAAGCGCTTCGGGGCTCTGCTCTCAGACCAGGGCTTCGACCTCATGAACAAGTTCCTGACCTACTTCCCCGGGAGGAGGATCAGCGCTGAGGACGGCCTCAAGCATGAGTATTTCCGCGAGACCCCCCTCCCCATCGACCCCTCCATGTTCCCCACGTGGCCCGCCAAGAGCGAGCAGCAGCGTGTGAAGCGGGGCACCAGCCCGAGGCCCCCTGAGGGAGGCCTGGGCTACAGCCAGCTGGGTGACGACGACCTGAAGGAGACGGGCTTCCACCTTACCACCACGAACCAGGGGGCCTCTGCCGCGGGCCCCGGCTTCAGCCTCAAGTTCTGAAGGTCAGAGTGGACCCCGTCATGGGGAGAACTCAGCCGGGACCACAGGCGTGGCTACTGCGGCTGGAGCTGCGATGAGACTCGGAACTCCTCGTCTTACTTTGTGCTCCATGTTTTGTTTTTGTATTTTGGTTTGTAAATTTGTAGAATTAAATCATTTTCCTTGTTGTGGAGGAAAGAGCTGTGTTTTCTCCGTGACTTGCCAGGGCATCTTCGGGTGCCCACGTGGGGCAGCACAAACCTCCACACACCCTCTCCCACTCTCGACACGCACGGGGCTGGCTGGGCCGTGATTTGGAAAGGAACTGGTGGGAGCCGGGTGGATTGTTTAATCTTCGGAGCTGGAGACCTGTTTCTGTGTTGGGATGAGCGATGCCCTCTTGCCCCAACCCACTCGTCCAGACCAGCCCTGTCCACACAGGCCCCCGGCCCCCAACCCCCAGCCCCAGCTGTGCCAGCAGACTCGACAGGTTTTTATACAAGGTTGTTGAGTTTTAAAATGTATTAAAATATTCTTCGAGGAAAGCTC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def print_first_n_sequences(fasta_file, n):\n",
    "    # Open the FASTA file using pysam.FastaFile\n",
    "    sequence = fasta_file.fetch(chrom, start, end).upper()\n",
    "    with pysam.FastaFile(fasta_file) as fasta:\n",
    "        # Get the list of sequence names (headers)\n",
    "        sequence_names = fasta.references\n",
    "\n",
    "        # Print the first n sequences\n",
    "        for i in range(min(n, len(sequence_names))):\n",
    "            sequence_name = sequence_names[i]\n",
    "            sequence = fasta.fetch(sequence_name)\n",
    "            print(f\"> {sequence_name}\\n{sequence}\\n\")\n",
    "\n",
    "print_first_n_sequences(fasta, 5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = fasta_file.fetch(\"110100000,Gene_ID:NCBI:728642,Refseq_ID:NM_001313896,Species:Homo\", 0, 100).upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATATCATCATTAGGCGTCAACACAGGAAGTGAGGATACTTCTGGCGAGCGCCGGTTGCTGTTTCTTCTCAGGCTCAGGGACCGGCCGCGGCCCCGTAGGG'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deeploc_torch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
